test/contracts/acl/acl_params.js

```diff
@@ -0,0 +1,206 @@
+
+        before(async () => {
+          overGasLimitOracle = await FailingOutOfGasOracle.new()
+          aclParams = permissionParamEqOracle(overGasLimitOracle.address)
+        })
+
+        testOraclePermissions({ shouldHavePermission: false })
+
+        describe('gas', () => {
+          describe('when permission is set for ANY_ADDR', () => {
+            // Note `evalParams()` is called twice when calling `hasPermission` for `ANY_ADDR`, so
+            // gas costs are much, much higher to compensate for the 63/64th rule on the second call
+            const MEDIUM_GAS = 3000000
+            const LOW_GAS = 2900000
+
+            beforeEach(async () => {
+              await acl.grantPermissionP(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, [aclParams])
+            })
+
+            it('ACL disallows and uses all gas when given large amount of gas', async () => {
+              assert.isFalse(await acl.hasPermission(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, { gas: MAX_GAS_AVAILABLE }))
+
+              const hasPermissionTxHash = await acl.hasPermission.sendTransaction(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, { gas: MAX_GAS_AVAILABLE })
+              const hasPermissionGasConsumed = web3.eth.getTransactionReceipt(hasPermissionTxHash).gasUsed
+              // Surprisingly, the actual gas used is quite a lot lower than expected, but it is
+              // unclear if this is a ganache issue or if there are gas refunds we're not taking
+              // into account
+              assert.closeTo(hasPermissionGasConsumed, getExpectedGas(MAX_GAS_AVAILABLE), 105000)
+            })
+
+            it('ACL disallows and uses all gas when given medium amount of gas', async () => {
+              assert.isFalse(await acl.hasPermission(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, { gas: MEDIUM_GAS }))
+
+              const hasPermissionTxHash = await acl.hasPermission.sendTransaction(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, { gas: MEDIUM_GAS })
+              const hasPermissionGasConsumed = web3.eth.getTransactionReceipt(hasPermissionTxHash).gasUsed
+              assert.closeTo(hasPermissionGasConsumed, getExpectedGas(MEDIUM_GAS), 10000)
+            })
+
+            it('ACL reverts when given small amount of gas', async () => {
+              await assertRevert(acl.hasPermission(ANY_ADDR, mockAppAddress, MOCK_APP_ROLE, { gas: LOW_GAS }))
+            })
+          })
+
+          describe('when permission is set for specific address', async () => {
+            const MEDIUM_GAS = 190000
+            // Note that these gas values are still quite high for causing reverts in "low gas"
+            // situations, as we incur some overhead with delegating into proxies and other checks.
+            // Assuming we incur 40-60k gas overhead for this, we only have ~140,000 gas left.
+            // After the oracle call, we only have 140,000 / 64 ~= 2000 gas left, which begins to
+            // quick run out with SLOADs.
+            const LOW_GAS = 180000
```

Comment:
This seems like quite a lot of gas, since I initially only assumed that we would be reverting under very low gas constraints.

However, in these tests, we are using a full organization, with proxies and all, and this adds a lot of overhead. We revert here because we run out of gas after the SLOAD for the `ANY_ADDR`'s permission params (if we flip the order that `hasPermission()` checks entities such that it checks the `ANY_ADDR` before the specific `who`, we can lower this `LOW_GAS` value).